### ISSUE TITLE
[BUG] Calculation inaccuracy caused by 'mlx.core.reshape' operation.

### ISSUE THREAD

**Issue by Redmept1on** [ml-explore/mlx#483]
**Describe the bug**
When I tried to construct a code fragment that called several mlx APIs consecutively, I found that in the result of the reshape operation, the number -0.31 was outputted as -0.310059 after the reshape operation, and this result indeed seemed to affect the accuracy of subsequent operations.

**To Reproduce**

Include code snippet
```python
import mlx.core as mx

a = mx.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5],dtype=mx.float16) 

new_shape = [3, 2]  
reshaped_array = mx.reshape(a, shape=new_shape)
print(reshaped_array)

repeated_array = mx.repeat(reshaped_array, repeats=2, axis=1)
print(repeated_array)

mean_result = mx.mean(repeated_array, axis=None, keepdims=True)
print(mean_result)
```
![image](https://github.com/ml-explore/mlx/assets/84373897/c4e71592-868a-416c-9201-36c0f616f31f)

**Expected behavior**
I implemented the same operation on the same array in numpy and found that the final calculated average value is more accurate.
```python
import numpy as np

a = np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=np.float16)

reshaped_a = a.reshape((3, 2))

repeated_a = np.repeat(reshaped_a, repeats=2, axis=1)

mean_value = np.mean(repeated_a)

print(reshaped_a)
print(repeated_a)
print(mean_value)
```
![image](https://github.com/ml-explore/mlx/assets/84373897/698760c2-a366-43b0-b9a9-ae3a68f993be)

**Desktop (please complete the following information):**
 - OS Version: 14.2.1
 - Version 0.7.0

**Additional context**
I think this may be due to a bug in 'mlx.core.reshape'.

**Comment by awni** [comment]
So if you just remove the reshape you get the same result, so I don't think it has anything to do with that:

```
a = mx.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=mx.float16)
repeated_array = mx.repeat(a, repeats=2)
mean_result = mx.mean(repeated_array, axis=None, keepdims=True)
print(mean_result)
```

Gives
```
array([0.03125], dtype=float16)
```

It looks like our reduction is slightly off 🤔 . I do not know if that's expected or a bug.  We do not do much to deal with numerically stable low precision reductions whereas it's possible NumPy and PyTorch do, but this is interesting...

Here is a simpler benchmark for all three:

```
import torch
import mlx.core as mx
import numpy as np

a = mx.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=mx.float16)
print(a.sum().item())

a = np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=np.float16)
print(a.sum().item())

a = torch.tensor([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=torch.float16)
print(a.sum().item())
```

Numpy and PyTorch are identical and our result is definitely quite a bit different:

```
0.19140625
0.18994140625
0.18994140625
```

Between CPU and GPU MLX gets the same result, so it doesn't seem likely to be specific to our GPU kernels.

**Comment by awni** [comment]
CC @angeloskath who wrote our reductions might have more insight on this one.

**Comment by Redmept1on** [comment]
Yep, this is very interesting. I also wrote another test.py and found that using mlx.core.reshape/ numpy.array.reshape twice, and then reshaping the array back to the initial shape. By comparing the initial array with the array that underwent two reshape operations, I found false. This seems to indicate that the reshape operation itself can bring about a change in data accuracy, but I think if this is not a bug, then, the printing of the mlx.core.reshape operation seems to be an enhancement point, as only -0.31 becomes -0.310059, while other floating-point numbers do not, , which seems a bit strange.
```python
import mlx.core as mx

a = mx.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5],dtype=mx.float16) 
  
a = mx.reshape(a, shape=[3,2])
print(a)

b = mx.reshape(a,shape=[1,6])
print(b)

array_equal = mx.array_equal(a,b)
print(array_equal)
```
![image](https://github.com/ml-explore/mlx/assets/84373897/0e5d1655-a086-4d81-a519-0621a62f91bb)

```python
import numpy as np

a = np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=np.float16)

reshaped_a = a.reshape((3, 2))
reshaped_b = reshaped_a.reshape(1,6)


print(reshaped_a)
print(reshaped_b)
print(np.array_equal(reshaped_a,reshaped_b))
```
![image](https://github.com/ml-explore/mlx/assets/84373897/f603816d-034c-4abd-9617-2eb3bc4aa38c)

**Comment by angeloskath** [comment]
I have a weirder one for you...

```python
a = np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=mx.float16)
alphas = np.split(a, len(a))

print(alphas)
# [array([-2.5], dtype=float16),
#  array([-1.5], dtype=float16),
#  array([-0.31], dtype=float16),
#  array([0.5], dtype=float16),
#  array([1.5], dtype=float16),
#  array([2.5], dtype=float16)]

print(sum(alphas))
# array([0.1914], dtype=float16)
```

Our cpu reduction kernel for float16 is literally a for loop

```c++
while (size-- > 0) {
  *accum += *x;
  x++;
}
```

**Comment by angeloskath** [comment]
Doing the summation in all possible orderings (all 720 of them) we get the following 4 different values together with their corresponding number of occurrences

```python
[(0.1875, 60), (0.189453125, 492), (0.19140625, 24), (0.18994140625, 144)]
```

Moreover quoting from the numpy docs

> For floating point numbers the numerical precision of sum (and np.add.reduce) is in general limited by directly adding each number individually to the result causing rounding errors in every step. However, often numpy will use a numerically better approach (partial pairwise summation) leading to improved precision in many use-cases. This improved precision is always provided when no axis is given.

So to sum up, we use a naive summation algorithm which has pretty bad roundoff errors. I think we can probably switch to partial pairwise summation without too much of a performance overhead as we can do it in each simd group (if the axis being summed is the fast moving axis). I would change this to an `enhancement` label or even better close and open a new issue with the title "Switch sum to partial pairwise summation".

Let me know what you guys think.

**Comment by angeloskath** [comment]
Also as expected (now that we read the docs)

```python
a = np.tile(np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=np.float16)[:, None], (1, 128))
print(a.sum(0))
array([0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914,
       0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914, 0.1914],
      dtype=float16)
```

**Comment by Redmept1on** [comment]
Based on your description, this issue should be an enhancement. In addition, I have another question. In the official MLX documentation, regarding the description of 'mlx. core. array_equal', it says that 'Returns True if and only if the arrays have the same shape and their values are equal.' But the program actually returns 'true' instead of 'True', in PyTorch or TensorFlow, this issue is considered as a mismatch between the output and the document or a typo related bug. I would like to know MLX's opinion on this type of issue.

**Comment by Redmept1on** [comment]
After thinking about this issue, I now feel that this should be listed as a bug related to mlx.core.float16, because specifically targeting the number -0.31, it becomes -0.310059 after printing, which is unexpected in itself. In addition, the document mlx.core.array emphasizes that 'The arrays need not have the same type to be considered equal.', However, when I run the following code, the result is false, which further indicates that mlx.core.float16 caused an unexpected error.If this data is applied to other downstream APIs, it may have a significant impact on the experimental results.
```python
import mlx.core as mx

a = mx.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5],dtype=mx.float16) 
print(a)
b = mx.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5],dtype=mx.float32)  
print(b)
c = mx.array_equal(a,b)
print(c)
```
![image](https://github.com/ml-explore/mlx/assets/84373897/6ebf9bdb-3e4c-4d86-b573-1e356d27c723)

**Comment by angeloskath** [comment]
I understand, however, the number `-0.31` is not representable in `float16`. All frameworks in general will represent it as `-0.31005859375`. It has nothing to do with MLX.

```python
a = np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=np.float16)
b = np.array([-2.5, -1.5, -0.31, 0.5, 1.5, 2.5], dtype=np.float32)
print(np.array_equal(a, b))
# also prints False

print(np.float16(-0.31).item())
# prints -0.31005859375
```

**Comment by awni** [comment]
@angeloskath thanks for finding the docs on NumPy, that makes total sense. I guess PyTorch / MPS must be doing the same as well.

I think we should add an issue for more numerically stable low precision reductions and mark it as an enhancement.

@Redmept1on 
> it says that 'Returns True if and only if the arrays have the same shape and their values are equal.' But the program actually returns 'true' instead of 'True',

The program returns an `mx.array` scalar boolean array which can be implicitly cast to a Python boolean. So it is correct in that respect. It does not print `True` and someone is working on better Python type formatting #354 to improve that.

>  'The arrays need not have the same type to be considered equal.',

They do not need the same type but they need to be the bitwise equivalent when cast to the same type. When comparing to slightly different floating point numbers `array_equal` is not the right thing. You should use `mx.allclose`.
